toolchain/check/convert.cpp

```diff
@@ -1208,19 +1208,60 @@ auto Convert(Context& context, SemIR::LocId loc_id, SemIR::InstId expr_id,
       if (!target.diagnose) {
         return context.emitter().BuildSuppressed();
       }
-      // TODO: Should this message change to say "object of type" when
-      // converting from a reference expression?
-      CARBON_DIAGNOSTIC(ImplicitAsConversionFailure, Error,
-                        "cannot implicitly convert value of type {0} to {1}",
-                        TypeOfInstId, SemIR::TypeId);
-      CARBON_DIAGNOSTIC(ExplicitAsConversionFailure, Error,
-                        "cannot convert value of type {0} to {1} with `as`",
-                        TypeOfInstId, SemIR::TypeId);
-      return context.emitter().Build(loc_id,
-                                     target.kind == ConversionTarget::ExplicitAs
-                                         ? ExplicitAsConversionFailure
-                                         : ImplicitAsConversionFailure,
-                                     expr_id, target.type_id);
+      bool explicit_as = target.kind == ConversionTarget::ExplicitAs;
+      auto type_of_expr_id = context.insts().Get(expr_id).type_id();
+      if (target.type_id == SemIR::TypeType::SingletonTypeId ||
+          sem_ir.types().Is<SemIR::FacetType>(target.type_id)) {
+        if (context.types().Is<SemIR::FacetType>(type_of_expr_id)) {
+          CARBON_DIAGNOSTIC(
+              ConversionFailureFacetToFacet, Error,
+              "cannot{0:| implicitly} convert type {1} that implements {2} to "
+              "type implementing {3}{0: with `as`|}",
+              BoolAsSelect, InstIdAsType, TypeOfInstId, SemIR::TypeId);
+          return context.emitter().Build(loc_id, ConversionFailureFacetToFacet,
+                                         explicit_as, expr_id, expr_id,
+                                         target.type_id);
+        } else if (auto facet_access_type =
+                       context.insts().TryGetAs<SemIR::FacetAccessType>(
+                           expr_id)) {
+          CARBON_DIAGNOSTIC(
+              ConversionFailureFacetAccessTypeToFacet, Error,
+              "cannot{0:| implicitly} convert type {1} that implements {2} to "
+              "type implementing {3}{0: with `as`|}",
+              BoolAsSelect, InstIdAsType, TypeOfInstId, SemIR::TypeId);
+          return context.emitter().Build(
+              loc_id, ConversionFailureFacetAccessTypeToFacet, explicit_as,
+              expr_id, facet_access_type->facet_value_inst_id, target.type_id);
+        } else if (type_of_expr_id == SemIR::TypeType::SingletonTypeId) {
+          CARBON_DIAGNOSTIC(
+              ConversionFailureTypeToFacet, Error,
+              "cannot{0:| implicitly} convert type {1} to type implementing "
+              "{2}{0: with `as`|}; type {1} does not implement the required "
+              "interface(s)",
```

Comment:
The `; type {1} does not implement the required interface(s)` part here is inconsistent with the prior diagnostic. Since they're diagnosing the same case with the only difference being whether the source type has locally-known constraints, it seems like the diagnostics ought to present the same information.

In discussion with @josh11b, we think the right place to produce this extra information about missing `impl`s would be to produce notes as part of impl lookup that describe exactly which impls are missing (as well as which impls we tried but failed to use and potentially why) -- though not as part of this PR. Once those notes are added, this extra information would be redundant, so let's omit it here:
```suggestion
              "{2}{0: with `as`|}",
```
(This diagnostic could then also be combined with the previous two with another select, but maybe it's not worthwhile.)

Reply:
Done.

I think combining them is a bit of a pain cuz it's more than just a bool, we also have to conditionally construct a FacetAccessType, so we have to also conditionally unwrap the optional, it just makes for a lot of conditions. That said, I can combine two other diagnostics, and I did combine these branches a little more.